include/ginkgo/core/base/array.hpp

```diff
@@ -294,14 +298,36 @@ class Array {
             this->resize_and_reset(0);
             return *this;
         }
-        this->resize_and_reset(other.get_num_elems());
-        exec_->copy_from(other.get_executor().get(), num_elems_,
+        // We allow resizing only if we are an array, we cannot resize a view!
+        // If we are a view, we do bound checking to ensure the data can be
+        // copied over.
+        if (data_.get_deleter().target_type() != typeid(view_deleter)) {
```

Comment:
Can you export this in a helper function? We do that at least twice, and it is not a trivial operation. A helper function would also explain what is done here. Also, if we find a better way to do it, we only have to change a single function.

The problem I see is that we allow the user to use a default deleter with a data pointer (Argument `DeleterType deleter`). That could potentially be a self-written view, and almost impossible (currently, I don't know how) to detect.
We can, however, do the opposite: check if the deleter is the `default_deleter` (which we would treat normally) and if not, treat it as a view (prevent allocations / `resize_and_reset`).

In case we change it, we might have to rewrite the documentation: instead of writing `a view`, we should probably write `a view or an array with a custom deleter` to document that behavior.
If we keep the current handling (checking against `view_deleter`), we might have to document that using a custom deleter is UB when doing assignments (when on the left side of assignments).